### PR TITLE
[WFLY-12190] EJB: EJB over HTTP discovery configuration

### PR DESCRIPTION
**Pull request by tadamski**
https://issues.redhat.com/browse/EAP7-1236
https://issues.redhat.com/browse/WFLY-12190

Depends on:
https://github.com/wildfly/wildfly/pull/13305

Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:
- [ ] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue(s)
- [ ] Pull Request contains description of the issue(s)
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Attached commits represent units of work and are properly formatted

For bigger changes, major and minor component upgrades make sure your PR also meets following requirements:
- [ ] Pull Request requires a change to the documentation
- [ ] Documentation have been updated accordingly
- [ ] Tests were added to cover changes

For new features ensure as well:
- [ ] Analysis was done
- [ ] Test Plan has been done
- [ ] Tests were verified in advance

If you are not an active contributor of the WildFly project you can request sponsorship by one of the members to help guide you through the process.

### CODE REVIEW

File: ejb3/src/main/resources/org/jboss/as/ejb3/subsystem/LocalDescriptions.properties
Comment (darranl) [inline ejb3/src/main/resources/org/jboss/as/ejb3/subsystem/LocalDescriptions.properties]: TODO

**Review by darranl — CHANGES_REQUESTED** [review]
The model description constants need defining.

However as on the analysis before we add a new server to server protocol we need to review the security implications.

**Comment by tadamski** [comment]
@darranl I have updated the description. I have also added the possibility of HTTP discovery from standalone client (in attached PRs). Regarding security, EJB over HTTP protocol is integrated with Elytron and the invocations are secured. Specifically, I had to provide Elytron configuration for the tests to pass (discovery invocations to succeed). I agree that it has to be double-checked though.

**Comment by darranl** [comment]
@tadamski I know the server side is already covered by Elytron security, the bit I am not sure about is the client side of these server to server connection.  In this case there are a few scenarios that may need some integration: -

- SSLContext
- Authentication Configuration 
- Identity Propagation using Delegated Credential

**Comment by tadamski** [comment]
@darranl OK, I will try to investigate those but I would probably need some help. OTOH server->server invocations are already working in WildFly (with static discovery) so we may need to double-check that those are fine as well.

**Comment by bstansberry** [comment]
@tadamski #12891 no longers includes the http-client part of this; it's just a bug fix update. The http-client part of this is now in a 1.1 release so this PR should include the upgrade to that.

**Review by pferraro — CHANGES_REQUESTED** [review]
This change requires a model version bump, as well as transformers and transformation tests.

**Comment by darranl** [comment]
@tadamski we haven't really followed up on this together, it looks like it has a merge conflict anyway but do we want to set something up so we can look at the security implications in more detail?

**Comment by tadamski** [comment]
@darranl I have to update this PR but this is mostly config stuff. The key PR is https://github.com/wildfly/wildfly-http-client/pull/29 on which this PR relies. I would be happy to discuss it.

**Comment by luck3y** [comment]
test, please ignore

**Comment by jamezp** [comment]
I've removed the missing-reqs label as there was an [agreement ](https://issues.redhat.com/browse/EAP7-1236?focusedCommentId=14105599&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14105599)the test development can be late on this feature.

**Comment by fl4via** [comment]
This PR depends on https://github.com/wildfly/wildfly/pull/13305

**Comment by tadamski** [comment]
@jamezp merged; failing test seem unrelated; this should be ready to merge

**Comment by zhantemirov** [comment]
I see all checks passing, do anything else prevent this PR from merging? Thanks

File: ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/jboss/as/ejb3/main/module.xml
Comment (bstansberry) [inline ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/jboss/as/ejb3/main/module.xml:122]: This is a duplicate.

File: ee/src/main/java/org/jboss/as/ee/metadata/EJBClientDescriptorMetaData.java
Comment (bstansberry) [inline ee/src/main/java/org/jboss/as/ee/metadata/EJBClientDescriptorMetaData.java:51]: This field should be final.  The ones above and below as well.

Comment (bstansberry) [reply ee/src/main/java/org/jboss/as/ee/metadata/EJBClientDescriptorMetaData.java:51]: Also, there is no code that modifies this list, so something is missing somewhere.

File: ee/src/main/java/org/jboss/as/ee/metadata/EJBClientDescriptorMetaData.java
Comment (bstansberry) [inline ee/src/main/java/org/jboss/as/ee/metadata/EJBClientDescriptorMetaData.java:357]: This should be final and presumably static.

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:190]: s/Remote profile/A remote profile/g
s/specificies/specifies

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:192]: s/either static or dynamic/either a static or a dynamic/g

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:194]: s/to connect/to connect to/g

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:197]: s/monitoring avaiable/monitoring the available/g

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:199]: "connect to based on the gathered data."

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:195]: Perhaps s/user/administrator  ?

AIUI the end user making the request isn't involved, so it's a user who administers the server config.

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:204]: s/allows/allows the administrator/g

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:205]: s/`module`/The `module`/g

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:215]: s/`remoting-ejb-receiver`/The `remoting-ejb-receiver`/g

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:216]: s/procol/protocol

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:226]: s/`remote-http-connection`/The `remote-http-connection`/g

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc:227]: s/procol/protocol

**Review by bstansberry — CHANGES_REQUESTED** [review]
I've made some comments. 

For this one to move forward we should get an approval from @darranl.

File: ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/jboss/as/ejb3/main/module.xml
Comment (tadamski) [reply ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/jboss/as/ejb3/main/module.xml]: fixed

File: ee/src/main/java/org/jboss/as/ee/metadata/EJBClientDescriptorMetaData.java
Comment (tadamski) [reply ee/src/main/java/org/jboss/as/ee/metadata/EJBClientDescriptorMetaData.java]: This was a bug. I have added the code that allows to configure http connections from jboss-ejb-client.xml file and added a test for that EjbOverHttpDescriptorTestCase

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (tadamski) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc]: fixed

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (tadamski) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc]: fixed

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (tadamski) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc]: fixed

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (tadamski) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc]: fixed

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (tadamski) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc]: fixed

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (tadamski) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc]: fixed

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (tadamski) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc]: fixed

File: ee/src/main/java/org/jboss/as/ee/metadata/EJBClientDescriptorMetaData.java
Comment (tadamski) [reply ee/src/main/java/org/jboss/as/ee/metadata/EJBClientDescriptorMetaData.java]: fixed

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (darranl) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc]: I think the main question I would like to ask is - is there sufficient test coverage of clients using authentication client configuration for a dynamically discovered destination?

As mentioned earlier we have aspects such as SSL and the credentials for authentication which should already be covered by the EJB client to verification that the behaviour is correct.  In this case I believe it will be preferable for the resolved configuration to be based on the discovered destination.

**Comment by bstansberry** [comment]
The issues I've raised have all been addressed, but the discussion with Darran remains open.

File: testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/ejb/http/EjbOverHttpWrongCredentialsTestCase.java
Comment (darranl) [inline testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/ejb/http/EjbOverHttpWrongCredentialsTestCase.java:1]: Missing header

**Comment by darranl** [comment]
I have one comment for testing that I don't think should hold up this PR so is something more general.  At the moment our testsuite is becoming very slow to execute so the pattern of Configure -> Deploy -> Single Test -> Undeploy -> Revert Configuration adds a lot of steps for one test.  Ideally if we can find ways a common deployment can be used for multiple scenarios we can execute multiple tests for a single pair or configure and pair of deploy tasks.  Ideally finding ways to test the success and failure scenarios under the same test case.

But in the case of this specific change I see we have some simple authentication client configuration where a single configuration is available but I would still like to see something where we test a configuration where the correct configuration is only resolved / resolvable after discovery.  I don't necessarily have an issue with that being in a follow up PR if we can make sure it happens.

The reason I am seeing this piece as important is because it will be an area of code we must maintain for backwards compatibility so once we release a Final version we will not be able to change the resolution logic without the risk of breaking something developed against the original behaviour.

File: testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/ejb/http/EjbOverHttpWrongCredentialsTestCase.java
Comment (tadamski) [reply testsuite/integration/multinode/src/test/java/org/jboss/as/test/multinode/ejb/http/EjbOverHttpWrongCredentialsTestCase.java:22]: fixed

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc
Comment (tadamski) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc]: The test coverage for the dynamically discovered destinations would require additional analysis. OTOH, dynamic discovery is main route of invocation also for remoting and this commit just adds this feature for HTTP. Currently EAP/WildFly already relies on dynamic invocation and if we decide that the testsuite have to be extended it should be for done a number of scenarios, possibly independently of this PR.

Comment (darranl) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc]: Ok, lets go with what we have here - if you are happy HTTP is just using the same pattern as the existing approaches we should be good.

Comment (tadamski) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/EJB3.adoc]: Hmm, but just to clarify cause I'm not sure if I understand correctly. Discovery provider and HttpEjbReceiver (which is used to perform invocation to discovered destination) (also other receivers) resolve their AuthenticationContexts independently:
https://github.com/wildfly/wildfly-http-client/blob/master/ejb/src/main/java/org/wildfly/httpclient/ejb/HttpEJBReceiver.java#L132

Should this be modified?

**Review by bstansberry — CHANGES_REQUESTED** [review]
@tadamski This needs an update to the transformers to the previous management API version:

https://github.com/wildfly/wildfly/blob/master/ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java#L289

Ejb3SubsystemUnitTestCase and Ejb3TransformersTestCase should cover it as well.

**Comment by bstansberry** [comment]
@tadamski https://issues.redhat.com/browse/WFLY-13878 is to track my last request.

**Comment by chengfang** [comment]
@bstansberry I'll update the transformer to handle the new element `remote-http-connection`

**Pull request merged**
